File: 376. Wiggle Subsequence/Python/Wiggle Subsequence.py
class Solution(object):
    def wiggleMaxLength(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if len(nums) <= 1:
            return len(nums)
        i = 1
        while i < len(nums) and nums[i] == nums[i - 1]:
            i += 1
        if i == len(nums):
            return 1
        is_increase = nums[0] > nums[i]
        result = 2
        prev = nums[i]
        for num in nums[i + 1:]:
            if (is_increase and num > prev) or (not is_increase and num < prev):
                result += 1
                is_increase = not is_increase
            prev = num
        return result

File: 376. Wiggle Subsequence/Java/Wiggle Subsequence.java
public class Solution {
    public int wiggleMaxLength(int[] nums) {
        if (nums.length <= 1) {
            return nums.length;
        }
        int i = 1;
        while (i < nums.length && nums[i] == nums[i - 1]) {
            i++;
        }
        if (i == nums.length) {
            return 1;
        }
        boolean isIncrease = nums[0] > nums[i];
        int result = 2;
        for (int j = i + 1; j < nums.length; j++) {
            if ((isIncrease && nums[j] > nums[j - 1]) || (!isIncrease && nums[j] < nums[j - 1])) {
                result++;
                isIncrease = !isIncrease;
            }
        }
        return result;
    }
}